samples/matter/common/src/bridge/ble_connectivity_manager.cpp

```diff
@@ -288,22 +305,44 @@ void BLEConnectivityManager::DiscoveryCompletedHandler(bt_gatt_dm *dm, void *con
 
 	discoveryResult = true;
 exit:
-	if (provider) {
-		if (!provider->IsInitiallyConnected()) {
-			/* Provider is not initalized it so we need to call the first connection callback */
-			provider->GetBLEBridgedDevice().mFirstConnectionCallback(
-				discoveryResult, provider->GetBLEBridgedDevice().mFirstConnectionCallbackContext);
-			provider->ConfirmInitialConnection();
-		}
-
-		VerifyOrReturn(CHIP_NO_ERROR == provider->NotifyReachableStatusChange(true),
-			       LOG_WRN("The device has not been notified about the status change."));
 
-		VerifyOrReturn(provider->ParseDiscoveredData(dm) == 0,
-			       LOG_ERR("Cannot parse the GATT discovered data."));
+	Platform::UniquePtr<DiscoveryHandlerCtx> discoveryCtx(Platform::New<DiscoveryHandlerCtx>());
+	if (!discoveryCtx) {
+		return;
 	}
 
-	bt_gatt_dm_data_release(dm);
+	discoveryCtx->mProvider = provider;
+	discoveryCtx->mDiscoveryData = dm;
+
+	CHIP_ERROR err = chip::DeviceLayer::PlatformMgr().ScheduleWork(
+		[](intptr_t context) {
+			Platform::UniquePtr<DiscoveryHandlerCtx> ctx(reinterpret_cast<DiscoveryHandlerCtx *>(context));
+			if (!ctx->mProvider->IsInitiallyConnected()) {
+				/* Provider is not initalized, so we need to call the first connection callback. */
+				CHIP_ERROR err = ctx->mProvider->GetBLEBridgedDevice().mFirstConnectionCallback(
+					ctx->mDiscoveryData,
+					ctx->mProvider->GetBLEBridgedDevice().mFirstConnectionCallbackContext);
+				ctx->mProvider->ConfirmInitialConnection();
+				VerifyOrReturn(CHIP_NO_ERROR == err, bt_gatt_dm_data_release(ctx->mDiscoveryData);
+					       Instance().RemoveBLEProvider(ctx->mProvider->GetBtAddress()););
+			}
+
+			if (CHIP_NO_ERROR != ctx->mProvider->NotifyReachableStatusChange(true)) {
+				LOG_WRN("The device has not been notified about the status change.");
+			}
+
+			if (0 != ctx->mProvider->ParseDiscoveredData(ctx->mDiscoveryData)) {
+				LOG_ERR("Cannot parse the GATT discovered data.");
+			}
+			bt_gatt_dm_data_release(ctx->mDiscoveryData);
```

Comment:
Shouldn't we delete `ctx` here?

Reply:
`ctx` is now an unique_ptr, so it will be automatically deleted.

Reply:
Ah ok, then it's ok.